### PR TITLE
Add new cases to cover rerror_policy on virtual disk

### PR DESCRIPTION
**Pull request by chunfuwen**
rerror_policy can have at least three values: ignore, report and stop
All of these scenarios are covered

Signed-off-by: chunfuwen <chwen@redhat.com>

### CODE REVIEW

**Comment by chunfuwen** [comment]
PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.rerror_policy_config.coldplug.set_rerror_policy_ignore.validate_guest_act
PASS 2-type_specific.io-github-autotest-libvirt.virtual_disks.rerror_policy_config.coldplug.set_rerror_policy_report.validate_guest_act
PASS 3-type_specific.io-github-autotest-libvirt.virtual_disks.rerror_policy_config.coldplug.set_rerror_policy_stop.validate_guest_act

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            LOG.error(str(e))
+            session.cmd("dmesg -C")
+            return False
+        else:
+            return error_msg in output
+
+    result = _check_dmeg_msg(error_msg)
+    if check_error_msg:
+        if not result:
+            test.fail("Failed to get expected message")
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:110]: Please print out the expected error message for debugging.

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                return False
+        except (remote.LoginError, virt_vm.VMError, aexpect.ShellError) as e:
+            LOG.error(str(e))
+            session.cmd("dmesg -C")
+            return False
+        else:
+            return error_msg in output
+
+    result = _check_dmeg_msg(error_msg)
+    if check_error_msg:
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:108]: The variable `check_error_msg`  is used for indicating that if the error message is expected to be in dmesg output per my understanding. So if I am right, I suggest the variable be named like 'error_msg_expected'

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        if not result:
+            test.fail("Failed to get expected message")
+    else:
+        if result:
+            test.fail("Unexpected messages are found")
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:113]: Please also include the unexpected message for debugging.

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    try:
+        session = vm.wait_for_login()
+        process.run("pidof qemu-nbd && killall qemu-nbd",
+                    ignore_status=True, shell=True)
+        # Execute read disk operation
+        cmd = "dd if=/dev/%s of=file" % device_target
+        session.cmd_status_output(cmd)
+        session.close()
+    except (remote.LoginError, virt_vm.VMError, aexpect.ShellError) as e:
+        logging.error(str(e))
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:67]: LOG

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    device_format = params.get("target_format")
+    device_target = params.get("target_dev")
+    rerror_policy_value = params.get("rerror_policy_value")
+    source_dict = {}
+    disk_src_dict = {}
+
+    # Prepare disk source xml
+    source_dict.update({"protocol": "nbd", "tls": "no"})
+    disk_src_dict.update({"attrs": source_dict})
+    disk_src_dict.update({"hosts": [{"name": nbd_server_host, "port": '10001'}]})
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:38]: I suggest `disk_src_dict` could be defined in cfg. That way could make this function shorter and clearer. Other users also could easily get to know what attributes are used for defining the disk.

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    :param params: dict wrapped with params
+    :param nbd_server_host: nbd server hostname
+    """
+    type_name = params.get("type_name")
+    disk_device = params.get("device_type")
+    device_bus = params.get("target_bus")
+    device_format = params.get("target_format")
+    device_target = params.get("target_dev")
+    rerror_policy_value = params.get("rerror_policy_value")
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:31]: BTW, if you adopt disk.setup_attrs(),  this function could be much shorter. However, it is just a suggestion and not a blocker.

**Comment by chunfuwen** [comment]
Add one more case here, new results are uploaded
******************************************************

PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.rerror_policy_config.coldplug.set_rerror_policy_ignore.validate_guest_act
PASS 2-type_specific.io-github-autotest-libvirt.virtual_disks.rerror_policy_config.coldplug.set_rerror_policy_ignore.start_guest
PASS 3-type_specific.io-github-autotest-libvirt.virtual_disks.rerror_policy_config.coldplug.set_rerror_policy_report.validate_guest_act
PASS 4-type_specific.io-github-autotest-libvirt.virtual_disks.rerror_policy_config.coldplug.set_rerror_policy_stop.validate_guest_act

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
+import logging
+import os
+import aexpect
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            process.run("pidof qemu-nbd && killall qemu-nbd",
+                        ignore_status=True, shell=True)
+            if not vm.pause():
+                test.fail("VM should be paused if rerror_policy='stop'")
+    finally:
+        try:
+            if nbd:
+                nbd.cleanup()
+        except Exception as ndbEx:
+            logging.info("Clean up nbd failed: %s" % str(ndbEx))
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 1 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:118]: The output debug information will be print out by previous statement
``
            LOG.debug("Disk operation in VM:\nexit code:\n%s\noutput:\n%s",
                      status, output)

```

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 2 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:116]: check_error_msg is boolean variable indicating whether it need check error message or not. And error message content is assigned to variable error_msg in previous statement.

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 3 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:121]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 4 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 5 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:38]: disk_src_dict need to be initialized dynamically due to nbd_server_host, which is the hostname of your server

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 6 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:31]: Thank you for reminding, are we really confident about   the way monolithic string was given to disk.setup_attrs(), is an approach we want to carry forward?

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+            vmxml.add_device(disk_xml_obj)
+            vmxml.sync()
+        vm.start()
+        vm.wait_for_login().close()
+    except virt_vm.VMStartError as e:
+        if status_error:
+            if hotplug:
+                test.fail("In hotplug scenario, VM should "
+                          "start successfully but not."
+                          "Error: %s", str(e))
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py]: update it by removing the whole block since it is not relevant now

**Comment by chunfuwen** [comment]
@qiankehan , please help take a look

**Review by qiankehan — APPROVED** [review]

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 6 above)
Comment (dzhengfy) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:31]: Yes, I confirm that this way is workable. See below simple sample I test with:
>>> disk_attrs = {'driver': {'rerror_policy': 'report'}, 'source': {'hosts': [{'protocol': 'rbd', 'name': '1.2.3', 'port': '1234'}]}}
>>> the_disk.setup_attrs(**disk_attrs)
>>> the_disk.fetch_attrs()
{'readonly': False, 'type_name': 'file', 'ready': False, 'transient': False, 'source': {'attrs': {}, 'hosts': [{'name': '1.2.3', 'port': '1234', 'protocol': 'rbd'}]}, 'share': False, 'mirror': False, 'driver': {'rerror_policy': 'report'}, 'drivermetadata': {'attrs': {'rerror_policy': 'report'}}}
>>> with open(the_disk.xml) as fp:
...     fp.read()
... 
'<?xml version=\'1.0\' encoding=\'UTF-8\'?>\n<disk type="file"><driver rerror_policy="report" /><source><host name="1.2.3" port="1234" protocol="rbd" /></source></disk>'

Per I know many people think this way is simple to use and make the codes less than before. So could you consider again?

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 5 above)
Comment (dzhengfy) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:38]: I know that. But you may consider to use '%s' in cfg, and then replace it with real value in python codes. I have already see some PRs using this way to handle dynamic situations. Could you help consider again?
Example:
```
in cfg:
disk_attrs = {'driver': {'rerror_policy': 'report'}, 'source': {'hosts': [{'protocol': 'rbd', 'name': '%s', 'port': '1234'}]}}

in python:
disk_attrs = disk_attrs % nbd_server_host
```

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 6 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:31]: @dzhengfy , could you give me some good reasons why monolithic hardcode string is the way to go forward?

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 5 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:38]: @dzhengfy , could you give me some good reasons why monolithic hardcode string is the way to go forward?

**Review by dzhengfy — APPROVED** [review]

File: libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py (hunk 5 above)
Comment (dzhengfy) [reply libvirt/tests/src/virtual_disks/virtual_disks_rerror_policy.py:38]: This is not a must to change. So it is not a blocker to merge.

**Pull request merged**